sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -510,40 +511,86 @@ case class AlterTableRenamePartitionCommand(
  *
  * The syntax of this command is:
  * {{{
- *   ALTER TABLE table DROP [IF EXISTS] PARTITION spec1[, PARTITION spec2, ...] [PURGE];
+ *   ALTER TABLE table DROP [IF EXISTS] PARTITION (spec1, expr1)
+ *     [, PARTITION (spec2, expr2), ...] [PURGE];
  * }}}
  */
 case class AlterTableDropPartitionCommand(
     tableName: TableIdentifier,
-    specs: Seq[TablePartitionSpec],
+    partitions: Seq[(TablePartitionSpec, Seq[Expression])],
     ifExists: Boolean,
     purge: Boolean,
     retainData: Boolean)
-  extends RunnableCommand {
+  extends RunnableCommand with PredicateHelper {
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
     val catalog = sparkSession.sessionState.catalog
     val table = catalog.getTableMetadata(tableName)
+    val resolver = sparkSession.sessionState.conf.resolver
     DDLUtils.verifyAlterTableType(catalog, table, isView = false)
     DDLUtils.verifyPartitionProviderIsHive(sparkSession, table, "ALTER TABLE DROP PARTITION")
 
-    val normalizedSpecs = specs.map { spec =>
-      PartitioningUtils.normalizePartitionSpec(
-        spec,
-        table.partitionColumnNames,
-        table.identifier.quotedString,
-        sparkSession.sessionState.conf.resolver)
+    val toDrop = partitions.flatMap { partition =>
+      if (partition._1.isEmpty && !partition._2.isEmpty) {
+        // There are only expressions in this drop condition.
+        extractFromPartitionFilter(partition._2, catalog, table, resolver)
+      } else if (!partition._1.isEmpty && partition._2.isEmpty) {
+        // There are only partitionSpecs in this drop condition.
+        extractFromPartitionSpec(partition._1, table, resolver)
+      } else if (!partition._1.isEmpty && !partition._2.isEmpty) {
+        // This drop condition has both partitionSpecs and expressions.
+        extractFromPartitionFilter(partition._2, catalog, table, resolver).intersect(
```

Comment:
I think this may be quite inefficient if we have a lot if partitions. What about converting the `partitionSpec` is `EqualsTo` expressions and add them as conditions? It would be great IMO if we can achieve this by enforcing in the syntax that we have either all `partitionSpec`s or all `expression`s. So if we have all `partition = value`, we have a `partitionSpec`, while if at least one is a comparison different from `=`, we have all `expression`s (including the `=`s). What do you think?

Reply:
Yeah, I agree. And the hard part may be how to convert a `partitionSpec` to an `EqualsTo`. 
I think it's better to let the `AstBuilder` to handle this. If so, we may have to have two `AlterTableDropPartitionCommand` instances in `ddl.scala`, one for all `partitionSpec` and one for all `expression`. But it maybe a bit weird.

Reply:
I mean how to define `AlterTableDropPartitionCommand` better in `ddl.scala`. need to handle both
`AlterTableDropPartitionCommand(
    tableName: TableIdentifier,
    partitions: Seq[Seq[Expression]],
    ifExists: Boolean,
    purge: Boolean,
    retainData: Boolean)`
and
`AlterTableDropPartitionCommand(
    tableName: TableIdentifier,
    partitions: Seq[TablePartitionSpec],
    ifExists: Boolean,
    purge: Boolean,
    retainData: Boolean)`
Maybe telling the different cases inside the method?

Reply:
I think we can (must) just have a single: `AlterTableDropPartitionCommand( tableName: TableIdentifier, partitionSpecs: Seq[TablePartitionSpec], partitionExprs: Seq[Seq[Expression]], ifExists: Boolean, purge: Boolean, retainData: Boolean)`. Indeed, we might have something like:
```
alter table foo drop partition (year=2017, month=12), partition(year=2018, month < 3);
```
where we have both a partition spec and an expression specification.

Reply:
hi, @mgaido91 there is one problem after I changed the syntax,
when i run sql `DROP PARTITION (p >=2)` it throws
`org.apache.spark.sql.AnalysisException: cannot resolve 'p' given input columns: []`
I'm trying to find a way to figure it out.

By the way, is a syntax like `((partitionVal (',' partitionVal)*) | (expression (',' expression)*))` legal? Because I wrote a antlr4 syntax test, but it didn't work as I supposed.

Besides, I was wrong that day. I think the if conditions won't be inefficient if there is a lot of partitions. it maybe inefficient if there are a lot of dropPartitionSpec which I don't think can happen easily.

Reply:
@DazhuangSu sorry I missed your last comment somehow.

Why do you say it would not be inefficient if you have a lot of partitions?I think it would be! Imagine that you partition per year and day. And you want to get the first 6 months of this year. The spec would be something like `(year = 2018, day < 2018-07-01)`. Imagine we have a 10 years history. With the current implementation, we would get back basically all the the partitions from the filter, ie. roughly 3.650 and then it will intersect those. Anyway, my understanding is that such a case would not even work properly, as it would try drop the intersect of:
```
Seq(Seq("year"-> "2018", "day" -> "2018-01-01", ...)).intersect(Seq(Map("year"->"2018")))
```
which would result in an empty Seq, so we would drop nothing. Moreover, I saw no test for this case in the tests. Can we add tests for this use case and can we add support for it if my understanding that it is not working is right? Thanks

Reply:
@mgaido91 I understand your point, yes it would be inefficient. I will work on this soon

Reply:
thank you @DazhuangSu